File: find-peak-element_162.py
#TimeComplexity: logN
#SpaceComplexity: O(1)
#It successfully runs on leetcode

class Solution:
    def findPeakElement(self, nums: List[int]) -> int:
        low = 0
        high = len(nums)-1
        
        while low <= high:
            mid=low+(high-low)//2
            if (mid==0 or nums[mid]>nums[mid-1]) and (mid==len(nums)-1 or nums[mid]>nums[mid+1]):
                return mid
            elif nums[mid] < nums[mid+1]:
                low=mid+1
            else:
                high=mid-1

File: Find Peak Element-162.java
class Solution {
    public int findPeakElement(int[] nums) {
// if array is empty
        if(nums==null || nums.length==0)
        {
            return -1;
        }

        int low =0;
        int high = nums.length-1;
        // Binary search
        while(low<=high)
        {
            int mid = low+(high-low)/2; // To prevent Integer Overflow Exception
            // If middle element is the largest amongst its neighbours 
            // Or if the mid element is first or last element the we comapre with -infinity so we return the element itself
            if((mid == low || nums[mid]>nums[mid-1]) && (mid==high ||nums[mid]>nums[mid+1]))
            {
               return mid;
            }
            // If the middle element is greater than element on its right side then move left
            else if (nums[mid]>nums[mid+1])
            {
                high = mid -1;
            }
             // If the middle element is greater than element on its left side then move right
            else {
                low = mid+1;
            }
        }
        return -1;
    }
}
